pkg/ccl/backupccl/restore_job.go

```diff
@@ -374,6 +374,12 @@ func WriteDescriptors(
 				return err
 			}
 			b.CPut(catalogkeys.EncodeNameKey(codec, desc), desc.GetID(), nil)
+
+			// We also have to put a system.namespace entry for the public schema
+			// if the database does not have a public schema backed by a descriptor.
+			if !desc.HasPublicSchemaWithDescriptor() {
+				b.CPut(catalogkeys.MakeSchemaNameKey(codec, desc.GetID(), tree.PublicSchema), keys.PublicSchemaID, nil)
```

Comment:
Naive question: is it safe to do this without a version gate?

Reply:
Yeah, we definitely always want to be inserting this namespace entry even in 21.2 - infact this should be backported.